### PR TITLE
Explicit TVD-RK2

### LINKED ISSUE #6600: The necessity to have a TVD time stepping scheme
### Description of the enhancement or error report
A second-order TVD (total variation diminishing) explicit time stepping scheme is needed in MOOSE.

Yidong will be working on the implementation of the TVD-RK2 scheme in MOOSE.

### Rationale for the enhancement or information for reproducing the error
It is well proved in theory and demonstrated that it is necessary to use a TVD stepping scheme in order to avoid oscillations at the presence of strong discontinuities for solving the Euler or Burgers' equations. 

Reference: see Figure 1 in the article
Gottlieb, S., & Shu, C. W. (1998). Total variation diminishing Runge-Kutta schemes. Mathematics of computation of the American Mathematical Society, 67(221), 73-85.
tvdrk

### Identified impact
Will add a new class in TimeIntegrators.
Will add new test cases in TimeIntegrators.
This TVD-RK2 time stepping scheme is a MUST for fully-stabilized, accuracy solution of conservation laws for strong discontinuity / strong shock phenomena, which is in addition to the the MUST of a fully-stabilized spatial discretization scheme. Direct applications are computational thermal-hydrodynamics, elasto-hydrodynamics, magneto-hydrodynamics.

### PR DESCRIPTION
**Pull request by yidongxiainl**
### Complete each of the following items before creating a PR

Added explicit TVD (total-variation-diminishing) 2nd-order time integration method
Added test cases for explicit TVD-RK2

See Eq. (2.4) in reference article:
Gottlieb, S., & Shu, C. W. (1998). Total variation diminishing Runge-Kutta schemes. Mathematics of computation of the American Mathematical Society, 67(221), 73-85.
[Cited 1002 times by Mar 28, 2016]

u^{(1)} = u^n + dt \* L(u^n)
u^{n+1} = 1/2 \* u^n + 1/2 \* u^{(1)} + 1/2 \* dt \* L(u^{(1)})

The main purpose of using explicit TVD-RK2 is to achieve non-oscillatory numerical solution for hyperbolic systems of equations. An example is shown in Figure 1.
tvdrk

Validation of implementation: Figure. L2-error vs. time step size (dt)
Obtained by successively reducing dt from 1e-4 to 5e-5, to 2.5e-5 
Temporal order of convergence is about 2.65, which validates the 2nd-order accuracy.
The super-convergence observed here is due to the choice of the test case
fig_l2_tvdrk2

closes #6600

### CODE REVIEW

**Comment by permcody** [comment]
:+1: @andrsd please review in @jwpeterson's absence

File: framework/src/executioners/Transient.C [hunk 1]
```diff
@@ -736,7 +736,8 @@ Transient::setupTimeIntegrator()
     case 2: ti_str = "CrankNicolson"; break;
     case 3: ti_str = "BDF2"; break;
     case 4: ti_str = "ExplicitMidpoint"; break;
-    case 5: ti_str = "LStableDirk2"; break;
+    case 5: ti_str = "ExplicitTVDRK2"; break;
+    case 6: ti_str = "LStableDirk2"; break;
```
Comment (andrsd) [inline framework/src/executioners/Transient.C]: `LStableDirk2` should remain `5` and the new one will be `6`.

File: test/tests/time_integrators/tvdrk2/1d-linear.i [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+[Variables]
+  [./u]
+    order = FIRST
+    family = LAGRANGE
+
+    [./InitialCondition]
+      type = FunctionIC
+      function = ic
+    [../]
```
Comment (andrsd) [inline test/tests/time_integrators/tvdrk2/1d-linear.i]: This needs to be in the top-level `[ICs]` block.

File: test/tests/time_integrators/tvdrk2/2d-quadratic.i [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+[Variables]
+  [./u]
+    order = SECOND
+    family = LAGRANGE
+
+    [./InitialCondition]
+      type = FunctionIC
+      function = ic
+    [../]
```
Comment (andrsd) [inline test/tests/time_integrators/tvdrk2/2d-quadratic.i]: This needs to be in the top-level `[ICs]` block.

File: test/tests/time_integrators/tvdrk2/2d-quadratic.i [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  [./ffn]
+    type = UserForcingFunction
+    variable = u
+    function = forcing_fn
+    implicit = false
+  [../]
+[]
+
+[BCs]
+  active = 'all'
```
Comment (andrsd) [inline test/tests/time_integrators/tvdrk2/2d-quadratic.i]: You can remove this.

File: framework/src/executioners/Transient.C (hunk 1 above)
Comment (andrsd) [inline framework/src/executioners/Transient.C]: And also line 52 in `Transient.C` is out of sync with this. It is missing the new `TVDRK2` scheme. I suggest to use `explicit-tvd-rk-2`. The dash between `rk` and `2` is to be consistent with our `rk-2` name.

Comment (yidongxiainl) [inline framework/src/executioners/Transient.C]: OK.

File: framework/src/timeintegrators/ExplicitTVDRK2.C [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    //
+    // where:
+    // .) M is the mass matrix.
+    // .) f(t_n+dt/2, Y_2) is the residual from stage 2.
+    // .) The minus sign is already baked in to the non-time
+    //    residuals, so it does not appear here.
+    // .) Although this is an update step, we have to do a "solve"
+    //    using the mass matrix.
+    residual.add(1.0, _Re_time);
+    residual.add(0.0, _residual_old);
```
Comment (jwpeterson) [inline framework/src/timeintegrators/ExplicitTVDRK2.C]: You're adding 0.0 times something... so just delete this line?

Comment (yidongxiainl) [inline framework/src/timeintegrators/ExplicitTVDRK2.C]: Alright, I see. Will fix.

Comment (yidongxiainl) [inline framework/src/timeintegrators/ExplicitTVDRK2.C]: All fixed. Can we merge now?

**Comment by jwpeterson** [comment]
Code looks good, but whenever we add a new TimeIntegrator, we like to see a convergence plot to confirm the theoretical order... can you post an image here that demonstrates this?  You don't have to change the tests or anything.

**Comment by yidongxiainl** [comment]
fig_l2_tvdrk2
Figure. L2-error vs. time step size (dt)
Obtained by successively reducing dt from 1e-4 to 5e-5, to 2.5e-5 
Temporal order of convergence is about 2.65, which validates the 2nd-order accuracy. The super-convergence observed here is due to the choice of the test case.

**Comment by yidongxiainl** [comment]
The convergence plot is appended in  the top post, along with the overall introduction. @jwpeterson

**Comment by jwpeterson** [comment]
Thanks!

**Pull request merged**